Provide cockpit/ws image for arm64

marvinruder opened the pull request:
This PR implements multi-platform images for the `ws` container as requested in #17569. It is inspired by both the implementation and review comments of #17935, but excludes changes in the documentation or unrelated scripts (I do support those changes, but assume it is better to handle those in a separate PR).

I wish to highlight the following changes:

> [!WARNING]  
> The section below is outdated.

<details>

- Multi-platform images are build by Docker Buildx for `linux/amd64` and `linux/arm64` by default
- ~Due to the differences of syntax of commands required to build multi-platform images between Podman and Docker, the script is no longer compatible with Podman. If support for Podman is required, I will have to rewrite some commands and this will likely result in a more complex script (also I never worked with Podman before and have no idea to what extent it supports multi-platform images). If support for Podman is no longer required, perhaps the `$RUNC` variable should also be removed.~ I went ahead and implemented support for both Podman and Docker in https://github.com/cockpit-project/cockpit/pull/21969/commits/202d22b415f3689495393a80e02f7a669e21f8f7.
- The interpreter has been changed to `bash` to support iterating through a comma-separated list of platforms
- Smoke tests run on every platform image and let the container print its own architecture to ensure that everything went as expected
- Docker must use the `containerd` image store in order to support storing multi-arch images locally, which is required for running them before pushing. When building locally, ensure that your Docker service uses the `containerd` image store. Also, QEMU must be installed for building and running images for different platforms
- The GitHub actions machine type is now `ubuntu-24.04-arm`. This is not strictly necessary, since the setup can create the intended result on every architecture, but aarch64 runners appear to be more performant at emulating x86 (~ 7 minutes) than x86 runners are at emulating aarch64 (timeout after 20 minutes)

</details>

Fixes #17569

jelly commented on .github/workflows/build-ws-container.yml:
```diff
@@ -8,14 +8,31 @@ on:
 jobs:
   build:
-    runs-on: ubuntu-latest
+    runs-on: ubuntu-24.04-arm
```
If GitHub provides two possible runners, why don't not we split it up in different `runs-on` profiles so we can benefit from native and not emulated performance.

marvinruder replied:
At the moment, there is no easy way to merge an image built on two different machines into one tag. You cannot push a version of an image for a certain architecture into an existing tag in a registry, that would overwrite the entire image. If we built the image on two different machines, we would have to use two different tags like `cockpit/ws:337-amd64` and `cockpit/ws:337-arm64`. Those would be two separate images and not one multi-platform image, which is rather inconvenient, as you have to specify your machine’s architecture when using the image and therefore run into problems e.g. when sharing the same Docker Compose configuration across multiple machines with different architectures.

marvinruder replied:
> The tasks container is already built that way: https://github.com/cockpit-project/cockpituous/blob/main/.github/workflows/build-tasks.yml. The two arm64/x86 builds are then "merged" via a manifest. This seems much nicer?

Oh nice, with `podman manifest` it seems doable. There also is an experimental [`docker manifest`](https://docs.docker.com/reference/cli/docker/manifest/) command, will check whether it works the same and update the PR later today.

jelly replied:
I think we would prefer using `podman manifest` so we can easily test it locally.

marvinruder replied:
> I think we would prefer using `podman manifest` so we can easily test it locally.

I will definitely prioritize Podman, just wanted to check if an equivalent setup with Docker is possible so we can keep the current dual setup with the `$RUNC` variable – or is this no longer necessary at all?

jelly replied:
@martinpitt I think we can drop the docker based workflow and go full podman as we already do this for the tasks container?

martinpitt replied:
I faintly remember that we had to build with docker because podman built images were not compatible with at least older docker images. If that's not an issue any more, sure!

marvinruder replied:
> I faintly remember that we had to build with docker because podman built images were not compatible with at least older docker images. If that's not an issue any more, sure!

Just tested running `ghcr.io/marvinruder/cockpit`, built by Podman, on Debian Bullseye running Docker 20 (very old), no issues there.

marvinruder commented:
I rewrote the GitHub Action as well as the `release.sh` script based on https://github.com/cockpit-project/cockpituous/blob/main/.github/workflows/build-tasks.yml:

- For the tasks container, apart from `ghcr.io/cockpit-project/tasks`, the separate registry repository `ghcr.io/cockpit-project/tasks-tmp` is used to store the temporary single-architecture images. Instead of having to create a separate registry repository, I used `podman save` to save the OCI image as tarball, used GitHub Actions artifacts to make them available in the `manifest` job and used `podman load` to load the tarball.
- When `release.sh` is called with one command-line argument, the script will treat this argument as file path and will output the image as a tarball under that file path. If no argument is given, the image is pushed to the registry.
- To test pushing the image, I made the image tag (and related arguments like the registry to log into) configurable, retaining the default `quay.io/cockpit/ws`. If desired, I can revert that part later.

The images built by GitHub Actions are available under [ghcr.io/marvinruder/cockpit](https://github.com/users/marvinruder/packages/container/package/cockpit), one successful workflow run can be found [here](https://github.com/marvinruder/cockpit/actions/runs/14979075743).

martinpitt commented on .github/workflows/build-ws-container.yml:
```diff
@@ -6,14 +6,25 @@ on:
+      matrix:
+        build:
+          - label: amd64
+            runner: ubuntu-24.04
+          - label: arm64
+            runner: ubuntu-24.04-arm
```
Not too long ago I struggled with moving all our hardcoded versions to -latest, as the former tend to get out of date. E.g. ubuntu-20.04 was dropped recently. It really pains me to see that reintroduced here, but alas, 
[ubuntu-latest-arm is not a thing](https://github.com/orgs/community/discussions/148648#discussioncomment-11858187). So *sigh* I suppose we need to live with this.

martinpitt commented on .github/workflows/build-ws-container.yml:
```diff
@@ -27,8 +38,43 @@ jobs:
+        env:
+          IMAGE_TAG: localhost/cockpit-ws:release-${{ matrix.build.label }}
+
+      - name: Save image to artifacts
+        uses: actions/upload-artifact@v4
+        with:
+          name: cockpit-ws-${{ matrix.build.label }}
+          path: cockpit-ws-${{ matrix.build.label }}.tar
```
This is really nice! Much better than having to upload the temporary container image to the registry.

martinpitt commented on .github/workflows/build-ws-container.yml:
```diff
@@ -27,8 +38,43 @@ jobs:
+  manifest:
+    needs: build
+
+    runs-on: ubuntu-24.04
```
But *this* one can hopefully at least run on -latest?

marvinruder replied:
Sure, if you prefer that, I will update.

martinpitt commented on .github/workflows/build-ws-container.yml:
```diff
@@ -27,8 +38,43 @@ jobs:
+    needs: build
+
+    runs-on: ubuntu-24.04
+    timeout-minutes: 5
```
This isn't entirely trivial, needs download, upload, and some local file I/O. Let's use 20?

martinpitt commented on .github/workflows/build-ws-container.yml:
```diff
@@ -6,14 +6,25 @@ on:
   workflow_dispatch:
 
+env:
+  tag: ${{ vars.REGISTRY_REPO || 'quay.io/cockpit/ws' }}
```
This is a bit confusing; "tag" sounds to me like ":latest". Maybe `repo` or `registry_repo` to match the vars?

martinpitt commented on .github/workflows/build-ws-container.yml:
```diff
@@ -27,8 +38,43 @@ jobs:
-      - name: Build and push ws container
-        run: containers/ws/release.sh
+      - name: Push :latest tag
+        run: podman manifest push ws "${{ env.tag }}:latest"
```
This looks wrong, but it's just a badly named `env.tag`. See above.

martinpitt commented on containers/ws/release.sh:
```diff
@@ -1,31 +1,36 @@
-        exit 1
-    }
+command -v podman || {
+    echo "ERROR: podman required" >&2
```
So docker can't build manifests? Not that anyone in our team cares, but is there a reason to drop docker in this script?

marvinruder replied:
I just noticed, in the end, this script got changed less than expected and the podman-only stuff is only executed in the `manifest` job in the workflow, so I can re-add the `$RUNC` variable.

(Yes, Docker can build manifests as well, but AFAIK the command syntax is different)

martinpitt commented on containers/ws/release.sh:
```diff
@@ -1,31 +1,36 @@
-$RUNC rm -f $name
+podman rm -f $name
 
-$RUNC tag quay.io/cockpit/ws:release quay.io/cockpit/ws:$TAG
-$RUNC tag quay.io/cockpit/ws:release quay.io/cockpit/ws:latest
```
This was accidentally dropped?

marvinruder replied:
No, this is intentional; the full image tag (including the registry repository, e.g.: `quay.io/cockpit/ws:release`) is now configurable using the environment variable `$IMAGE_TAG`, because the GitHub Action workflow needs a different tag on different architectures (`localhost/cockpit-ws:release-${{ matrix.build.label }}`). Also the script will no longer build the final image that will be pushed to the registry, just one part of it for the current architecture.

Speaking of it, if the script is executed locally, one probably doesn’t always want to push it. I should remove the `podman push` command, it is not used by the GitHub Action workflow, and if someone still wants to push it from their local machine, it might be best to have them add the desired tags manually after the script built the image and stored it locally.

martinpitt replied:
Right, it makes sense to take out the push. It's useful to locally build and test a container, so that should do the tagging (variables are fine), but not the pushing.

martinpitt reviewed (changes requested):
Sorry for the delay! (I'm not working in the cockpit team this quarter). This looks really nice, I just have some questions/small stuff. Thanks for the comprehensive testing!

Can you please squash your commits and add a proper commit message?

Cheers! :star:

martinpitt commented:
@marvinruder :

>     * Update the build image to Fedora 42

Yes, agreed. But this is a bit finicky because reasons, and requires some more changes and coordination. I added it to our [pilot board](https://github.com/orgs/cockpit-project/projects/4?pane=issue&itemId=111010178), let's do that after landing this PR.

>     * Run the workflow more often (e.g. daily), but skip pushing to the registry if it already holds an image with the same version, with a possible override flag `FORCE_PUSH` as a workflow input. This way, new versions are available faster after a release, no unnecessary images are pushed every week when there is no update, and corrupted images can still be overwritten manually using the same tag if something goes wrong.

Yes, that makes sense. I slightly prefer if that goes into a separate PR, but if you want to move it all in one go, no objection. But make it a separate commit please.

Thanks!

marvinruder commented on .github/workflows/build-ws-container.yml:
```diff
@@ -6,14 +6,25 @@ on:
-    timeout-minutes: 20
-    env:
-      RUNC: docker
+    timeout-minutes: 10
```
Probably took it from https://github.com/cockpit-project/cockpituous/blob/main/.github/workflows/build-tasks.yml, but sure, I can change it back.

marvinruder commented:
> I slightly prefer if that goes into a separate PR

Agreed. Will create one after this one is merged.

@martinpitt I added my changes addressing your review comments in https://github.com/cockpit-project/cockpit/pull/21969/commits/6adf67b848f17a28feaaf478248ebb5793cd3463 if you want to have an isolated look at them before I squash everything.

martinpitt commented on containers/ws/release.sh:
```diff
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/bin/bash
```
No need to change to bash, all looks POSIXy here. But not important.

marvinruder replied:
Ah yes, that was only necessary in one of my QEMU-related drafts for iterating over comma-separated architectures.

martinpitt reviewed (approved):
Thank you, great work! I'll take this for a spin.

martinpitt commented:
@marvinruder I [ran the workflow](https://github.com/cockpit-project/cockpit/actions/runs/15129993898/job/42528949947) and it failed: the [build step tried to push the image](https://github.com/cockpit-project/cockpit/actions/runs/15129993898/job/42528949947#step:4:908) already, which is premature -- that should happen in the manifest step. Do you want to look into that?

marvinruder commented:
> @marvinruder I [ran the workflow](https://github.com/cockpit-project/cockpit/actions/runs/15129993898/job/42528949947) and it failed: the [build step tried to push the image](https://github.com/cockpit-project/cockpit/actions/runs/15129993898/job/42528949947#step:4:908) already, which is premature -- that should happen in the manifest step. Do you want to look into that?

@martinpitt Due to https://github.com/cockpit-project/cockpit/blob/c26c92ffaf59ccf9586dcf124c16e853e9dfbfe3/.github/workflows/build-ws-container.yml#L35-L39 the workflow uses the old `release.sh`. If you don’t want to create a new release immediately, a simple git tag `tmp` or similar on the current commit on `main` should help.

martinpitt commented:
Oh right, thanks! Let's retry after the next release then.

Venefilyn commented:
> Just pulled the image published by [cockpit-project/cockpit/actions/runs/15168922376](https://github.com/cockpit-project/cockpit/actions/runs/15168922376) on my arm64 server, it works great and feels a lot faster, as emulation is no longer required.

Great to hear! Now time for us to do RISC-V next.. for science!
